### PR TITLE
Remove unnecessary dtype checks for complex types & disable complex dispatch for CPU min/max pointwise ops

### LINKED ISSUE #50064: Complex dispatch should be disabled on min/max functions
As of #36377, min/max functions were disabled for complex inputs (via dtype checks), however, min/max kernels are still compiled and dispatched for complex, see e.g.  
https://github.com/pytorch/pytorch/blob/e44b2b72bd4ccecf9c2f6c18d09c11eff446b5a3/aten/src/ATen/native/cpu/ReduceAllOpsKernel.cpp#L77
This dispatch should be disabled, and we should rely on errors produced by dispatch macro to not run those ops on complex, instead of doing redundant dtype checks. 

cc @ezyang @anjali411 @dylanbespalko @mruberry 
cc @t-vi, thanks for reporting!

### PR DESCRIPTION
**Pull request by imaginary-person**
Fixes #50064

**PROBLEM DESCRIPTION:**
1. Had not removed dtype checks for complex types in the previous PR (#50347) for this issue.
These type-checks were added in #36377, but are no longer necessary,
as we now rely upon dispatch macros to produce error messages.
2. dtype checks in `clamp_max()` and `clamp_min()` for complex inputs had not been removed either.
3. For min/max pointwise ops in TensorCompareKernel.cpp, complex dispatch had not been removed for min/max functions.

### **FIX DESCRIPTION:**
**FIX SUMMARY:**
1. Removed dtype checks added in #36377, and added 3 more in TensorCompare.cpp.
2. Removed dtype checks for complex inputs in `clamp_max()` and `clamp_min()`.
3.  Disabled complex dispatch for min/max pointwise ops in TensorCompareKernel.cpp.
4. Error messages in the exceptions raised due to min/max ops not being implemented are now checked for containing the text _not support_ (which can also be present in _not supported_), or _not implemented_, so one of them should be a part of error messages, in order for them to be informative.

**REASON FOR NOT CHANGING DISPATCH FOR CUDA AND CLAMP OPS**:

As for the CUDA min/max operations, their kernels do not seem to be compiled & dispatched for complex types anyway, so no further changes seem to be required. Basically, the dispatch macros currently being used don't have cases for complex types.

For example, 

1. the reduce CUDA ops use [AT_DISPATCH_ALL_TYPES_AND2](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/Dispatch.h#L548-L575) in [ReduceMinMaxKernel.cu](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/ReduceMinMaxKernel.cu), and that macro doesn't allow complex types.

2. In [MinMaxElementwiseKernel.cu](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/MaxMinElementwiseKernel.cu), the CUDA pointwise ops use [`AT_DISPATCH_FLOATING_TYPES_AND2`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/Dispatch.h#L240-L263) for non-integral & non-boolean types, and this marco doesn't have a case for complex types either.

3. [clamp CUDA ops](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/UnaryOpsKernel.cu#L170-L211) use `AT_DISPATCH_ALL_TYPES_AND2`, which doesn't have a case for complex types.

Similarly, [CPU clamp min/max ops](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp#L428-L458) use the `AT_DISPATCH_ALL_TYPES_AND `dispatch macro, which doesn't have a case for complex types.

**REASON FOR ADDING 3 dtype CHECKS:**
There are a few cases in which the methods corresponding to `min_stub()` or `max_stub()` are not called, so dispatch macros don't get invoked, resulting in no exceptions being raised. Hence, `dtype` checks are necessary at 3 places to raise exceptions:

1. https://github.com/pytorch/pytorch/blob/52dcc7299925de055d330781d2fe0dad71182829/aten/src/ATen/native/TensorCompare.cpp#L342
2. https://github.com/pytorch/pytorch/blob/52dcc7299925de055d330781d2fe0dad71182829/aten/src/ATen/native/TensorCompare.cpp#L422
3. https://github.com/pytorch/pytorch/blob/52dcc7299925de055d330781d2fe0dad71182829/aten/src/ATen/native/TensorCompare.cpp#L389

The first dtype check requirement can be verified from the following example Python code based on `test_complex_unsupported()`:
```
import unittest 
import torch

class MyTestCase(unittest.TestCase): 
  
   def test_1(self):
      t = torch.tensor((1 + 1j), device='cpu', dtype=torch.complex128) 
      with self.assertRaises(Exception): 
         torch.max(t, dim=0)
  
if __name__ == '__main__':  
    unittest.main()
```

### CODE REVIEW

**Comment by imaginary-person** [comment]
> ##  CI failures summary and remediations
> As of commit [cef471a](https://github.com/pytorch/pytorch/commit/cef471a065b33bc7ea4deb3fe3893ad4e7026c43) (more details [on the Dr. CI page](https://our.intern.facebook.com/intern/opensource/dr_ci/commit/167825833786582/cef471a065b33bc7ea4deb3fe3893ad4e7026c43/)):
> 
> * **2/2** failures introduced in this PR
> 
> ###  2 new failures recognized by patterns
> The following CI failures do not appear to be due to upstream breakages:
> 
> #### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/10138329) [pytorch_linux_xenial_py3_6_gcc5_4_test](https://circleci.com/gh/pytorch/pytorch/10138329?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) (1/2)
> **Step:** "Run tests" ([full log](https://our.intern.facebook.com/intern/opensource/ci/job/log/721443325404465/) | [diagnosis details](https://our.intern.facebook.com/intern/opensource/ci/job/721443325404465/) | [repeat rerun](https://our.intern.facebook.com/intern/opensource/ci/restart/721443325404465/))
> 
>  `Jan 13 19:17:10 AssertionError: RuntimeError not raised`
> #### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/10138002) [pytorch_linux_bionic_py3_6_clang9_test](https://circleci.com/gh/pytorch/pytorch/10138002?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) (2/2)
> **Step:** "Run tests" ([full log](https://our.intern.facebook.com/intern/opensource/ci/job/log/2733837303537796/) | [diagnosis details](https://our.intern.facebook.com/intern/opensource/ci/job/2733837303537796/) | [repeat rerun](https://our.intern.facebook.com/intern/opensource/ci/restart/2733837303537796/))
> 
>  `Jan 13 19:11:29 AssertionError: RuntimeError not raised`
> This comment was automatically generated by [Dr. CI](https://code.facebook.com/ci/dr-ci-info/) (expand for details).

@pytorchbot retest this please

**Comment by imaginary-person** [comment]
All the errors are related to test_complex_unsupported(). `AT_ERROR()` also uses `TORCH_CHECK()` to raise a RuntimeError,
so I'll check why it's not working as intended & will fix the bug after testing `test_complex_unsupported()` locally as soon as my local debug build would finish.

**Comment by imaginary-person** [comment]
@pytorchbot retest this please

> All the errors are related to` test_complex_unsupported()`. `AT_ERROR()` also uses `TORCH_CHECK()` to raise a RuntimeError,
> so I'll check why it's not working as intended & will fix the bug after testing `test_complex_unsupported()` locally as soon as my local debug build would finish.

I used DEBUG=1 to build but `python3 -m torch.utils.collect_env `shows `DEBUG: False`.
I'm building again.

**Comment by imaginary-person** [comment]
Found the bug. Tested locally.

It's worth noting that dispatch handlers are invoked in the methods registered for `min_stub` & `max_stub`.
However, the tests fail in those conditional cases in which these stubs are not being called!

We can't rely upon dispatch macros to raise exceptions if they're never invoked in the first place!
So, in these corner cases, we need dtype checks.

File: aten/src/ATen/native/TensorCompare.cpp [hunk 1]
```diff
@@ -340,6 +340,7 @@ static std::tuple<Tensor &,Tensor &> max_out_impl(Tensor& max, Tensor& max_indic
               max_indices.device(), " for indices output");
   dim = maybe_wrap_dim(dim, self.dim());
   if (_dimreduce_return_trivial_no_ident(max, self, dim, keepdim, "max")) {
+    TORCH_CHECK(!self.is_complex(), "max does not support complex inputs.");
```
Comment (imaginary-person) [inline aten/src/ATen/native/TensorCompare.cpp:343]: It's worth noting that dispatch handlers are invoked in the methods registered for `min_stub` & `max_stub`.
However, the tests fail in some conditional cases such as this one, in which these stubs are not being called!

We can't rely upon dispatch macros to raise exceptions if they're never invoked in the first place!
So, in these corner cases, we need dtype checks.

Comment (ngimel) [reply aten/src/ATen/native/TensorCompare.cpp:343]: Great, thanks for investigating!

**Comment by imaginary-person** [comment]
Build failure cause on Jenkins for pytorch-linux-bionic-rocm3.10-py3.6 is unrelated to the code in this PR because previous builds on other PRs also failed with the same cause:

```
test_kernel_cache_multi_gpu (__main__.TestTEFuser) ... Memory exception on virtual address 0x7fc601200000, node id 3 : Page not present
RuntimeError: test_jit_fuser_te failed! Received signal: SIGIOT
```

**Comment by imaginary-person** [comment]
> ## 💊 CI failures summary and remediations
> As of commit [52dcc72](https://github.com/pytorch/pytorch/commit/52dcc7299925de055d330781d2fe0dad71182829) (more details [on the Dr. CI page](https://our.intern.facebook.com/intern/opensource/dr_ci/commit/167825833786582/52dcc7299925de055d330781d2fe0dad71182829/)):
> 
> * **1/1** failures possibly[*](https://code.facebook.com/ci/dr-ci-info/) introduced in this PR
>   
>   * **1/1** non-CircleCI failure(s)
> 
> ### ci.pytorch.org: 1 failed
> * **Failed:** [`pr/pytorch-linux-bionic-rocm3.10-py3.6`](https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm3.10-py3.6-trigger/3453/)
> 
> This comment was automatically generated by [Dr. CI](https://code.facebook.com/ci/dr-ci-info/) (expand for details).

This failure is not due to the code in this PR, as Jenkins builds for other PRs are still failing due to the same cause.
For example, https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm3.10-py3.6-test2/2859/console

**Comment by ngimel** [comment]
This is looking really good! Can you please also remove dtype checks for clamp in UnaryOps.cpp, and rely on dispatch throwing the errors? Dispatch is not enabled for compelx currently, so as far as I can see no dispatch changes are necessary.

**Comment by imaginary-person** [comment]
> This is looking really good! Can you please also remove dtype checks for clamp in UnaryOps.cpp, and rely on dispatch throwing the errors? Dispatch is not enabled for compelx currently, so as far as I can see no dispatch changes are necessary.

Thanks for your advice! I'll do that right away.

**Review by ngimel — APPROVED** [review]
Let's wait for CI, otherwise looks good.

**Comment by imaginary-person** [comment]
Hello @ngimel, it seems that the Codecov/patch check is insignificant because it's related to a line of code that does a dtype check for complex types if a particular conditional statement is true, as no dispatch macro is invoked for that condition.

The Codecov/patch check failed because no test cases check for that condition, so that line of code was never visited at runtime.

**Comment by imaginary-person** [comment]
> > ## 💊 CI failures summary and remediations
> > As of commit [52dcc72](https://github.com/pytorch/pytorch/commit/52dcc7299925de055d330781d2fe0dad71182829) (more details [on the Dr. CI page](https://our.intern.facebook.com/intern/opensource/dr_ci/commit/167825833786582/52dcc7299925de055d330781d2fe0dad71182829/)):
> > 
> > * **1/1** failures possibly[*](https://code.facebook.com/ci/dr-ci-info/) introduced in this PR
> >   
> >   * **1/1** non-CircleCI failure(s)
> > 
> > ### ci.pytorch.org: 1 failed
> > 
> > * **Failed:** [`pr/pytorch-linux-bionic-rocm3.10-py3.6`](https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm3.10-py3.6-trigger/3453/)
> > 
> > This comment was automatically generated by [Dr. CI](https://code.facebook.com/ci/dr-ci-info/) (expand for details).
> 
> This failure is not due to the code in this PR, as Jenkins builds for other PRs are still failing due to the same cause.
> For example, https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm3.10-py3.6-trigger/3500/

This unrelated issue was fixed in #50518

**Comment by imaginary-person** [comment]
> Let's wait for CI, otherwise looks good.

Hello @ngimel, please let me know if I should add a test pertaining to `torch._aminmax()` for complex inputs in order to pass the Codecov/patch check. Thanks!

**Comment by ngimel** [comment]
Thanks for adding the test!

**Comment by imaginary-person** [comment]
> Thanks for adding the test!

Thanks for the opportunity to contribute, @ngimel!
Codecov/patch is passing now, but some other internal FB tests still seem to be failing.
I'm unable to see the failures, but if required, I'll make any necessary changes. Thanks!

**Comment by ngimel** [comment]
Internal failures are unrelated, I'm trying to land now. Thanks for the contribution!

**Pull request closed without merging**